src/libs/kata-types/src/config/hypervisor/mod.rs

```diff
@@ -0,0 +1,1061 @@
+    #[serde(default)]
+    pub enable_iommu: bool,
+
+    /// Enable IOMMU_PLATFORM, default false
+    ///
+    /// Enabling this will result in the VM device having iommu_platform=on set
+    #[serde(default)]
+    pub enable_iommu_platform: bool,
+}
+
+impl DeviceInfo {
+    /// Adjust the configuration information after loading from configuration file.
+    pub fn adjust_configuration(&mut self) -> Result<()> {
+        if self.default_bridges > 5 {
+            self.default_bridges = 5;
+        }
+
+        Ok(())
+    }
+
+    /// Validate the configuration information.
+    pub fn validate(&self) -> Result<()> {
+        if self.default_bridges > 5 {
+            return Err(eother!(
+                "The configured PCI bridges {} is too big",
+                self.default_bridges
+            ));
+        }
+        Ok(())
+    }
+}
+
+/// Configuration information for virtual machine.
+#[derive(Debug, Default, Deserialize, Serialize)]
+pub struct MachineInfo {
+    /// Virtual machine model/type.
+    #[serde(default)]
+    pub machine_type: String,
+
+    /// Machine accelerators.
+    /// Comma-separated list of machine accelerators to pass to the hypervisor.
+    /// For example, `machine_accelerators = "nosmm,nosmbus,nosata,nopit,static-prt,nofw"`
+    #[serde(default)]
+    pub machine_accelerators: String,
+
+    /// Add flash image file to VM.
+    ///
+    /// The arguments of it should be in format of ["/path/to/flash0.img", "/path/to/flash1.img"].
+    #[serde(default)]
+    pub pflashes: Vec<String>,
+
+    /// Default entropy source.
+    /// The path to a host source of entropy (including a real hardware RNG).
+    /// `/dev/urandom` and `/dev/random` are two main options. Be aware that `/dev/random` is a
+    /// blocking source of entropy.  If the host runs out of entropy, the VMs boot time will
+    /// increase leading to get startup timeouts. The source of entropy `/dev/urandom` is
+    /// non-blocking and provides a generally acceptable source of entropy. It should work well
+    /// for pretty much all practical purposes.
+    #[serde(default)]
+    pub entropy_source: String,
+
+    /// List of valid annotations values for entropy_source.
+    /// The default if not set is empty (all annotations rejected.)
+    #[serde(default)]
+    pub valid_entropy_sources: Vec<String>,
+}
+
+impl MachineInfo {
+    /// Adjust the configuration information after loading from configuration file.
+    pub fn adjust_configuration(&mut self) -> Result<()> {
+        let accelerators: Vec<&str> = self
+            .machine_accelerators
+            .split(',')
+            .map(|v| v.trim())
+            .collect();
+        self.machine_accelerators = accelerators.join(",");
+
+        for pflash in self.pflashes.iter_mut() {
+            resolve_path!(*pflash, "Flash image file {} is invalide: {}")?;
+        }
+        resolve_path!(self.entropy_source, "Entropy source {} is invalid: {}")?;
+
```

Comment:
Doesn't this only belong into `validate`?

Reply:
`adjust_configuration` is responsible for adjusting value,` validate` use another macro `validate_path`